### PR TITLE
Support deployment to AWS EMR

### PR DESCRIPTION
**Pull request by lijinf2**
Got all notebooks (spark-compat, pca, linear-regression, knn, and kmeans) tested and passed in AWS EMR

### CODE REVIEW

File: notebooks/aws-emr/README.md
Comment (leewyang) [inline notebooks/aws-emr/README.md:39]: Remove blank line.

File: notebooks/aws-emr/README.md
Comment (leewyang) [inline notebooks/aws-emr/README.md:55]: Might be nice to explain how MASTER, CORE, TASK map to the various spark nodes (in case folks want to adjust these in the future).  For instance, I'm wondering why we need three GPU nodes if the driver wasn't using GPU?

**Review by leewyang — COMMENTED** [review]
Minor comments, otherwise LGTM

File: notebooks/aws-emr/README.md
Comment (eordentlich) [inline notebooks/aws-emr/README.md:15]: gcs -> s3

File: notebooks/aws-emr/README.md
Comment (eordentlich) [inline notebooks/aws-emr/README.md:50]: This could be cpu node.   m4.2xlarge.

File: notebooks/aws-emr/README.md
Comment (eordentlich) [inline notebooks/aws-emr/README.md:52]: Can we have one of these with InstanceCount=2?

File: notebooks/aws-emr/init-configurations.json
Comment (eordentlich) [inline notebooks/aws-emr/init-configurations.json:76]: Interesting.  Do we have this elsewhere?   What happens if it is not set?

File: notebooks/kmeans.ipynb
Comment (eordentlich) [inline notebooks/kmeans.ipynb:60]: np.float32 -> dtype ?

File: notebooks/aws-emr/README.md
Comment (lijinf2) [reply notebooks/aws-emr/README.md:53]: Good suggestion. Just added a few words to explain why three GPU instances to get two workers.

File: notebooks/aws-emr/init-configurations.json
Comment (lijinf2) [reply notebooks/aws-emr/init-configurations.json:76]: Yeah, this also appears in Dataproce. 

It not set, there will be a permission error writing to /home/.cupy.

File: notebooks/aws-emr/README.md
Comment (leewyang) [reply notebooks/aws-emr/README.md:53]: So EMR is _randomly_ picking either a CORE or a TASK node to run JupyterLab?  I would assume that TASK nodes would be reserved for the executors, so JupyerLab would run on CORE (which could theoretically be a CPU machine).

Comment (lijinf2) [reply notebooks/aws-emr/README.md:53]: Yeah, printed out the node can be CORE or TASK. There could be a way to specifically assign the node.

File: notebooks/aws-emr/init-configurations.json
Comment (eordentlich) [reply notebooks/aws-emr/init-configurations.json:76]: We'll have to look into whether isolation is needed for this directory in the case of concurrent spark jobs in the same cluster.

File: notebooks/aws-emr/README.md
Comment (eordentlich) [reply notebooks/aws-emr/README.md:53]: CORE nodes run hdfs while TASK nodes do not.   That seems to be the main distinction and it doesn't constrain where the driver goes.   I guess we can force the driver to land on a cpu only node by giving that node more memory than the other nodes and setting the driver memory such that yarn must schedule it to that node.   The executors can then land on the gpu nodes based on the gpu resource requirements specified for executors.

Comment (eordentlich) [reply notebooks/aws-emr/README.md:53]: Ideally yarn should be 'smart' and not schedule a driver that doesn't request gpu resource to a gpu node when adequate resources are available on a cpu node.

File: notebooks/aws-emr/init-configurations.json
Comment (eordentlich) [inline notebooks/aws-emr/init-configurations.json:47]: Can you check formatting on this file?  (tabs vs spaces, etc)?  Looks like entries are misaligned.

**Comment by eordentlich** [comment]
@lijinf2 Can you address remaining few comments?   **No need to pass build** once it gets an approval to merge since no changes go through ci/cd testing.

File: notebooks/aws-emr/README.md
Comment (lijinf2) [reply notebooks/aws-emr/README.md:52]: (1) The cost for large memory can be higher than small memory + GPU. AWS pricing shows m4.4xlarge (64G) costs 0.8/h, and g4dn.2xlarge (32G + GPU) costs  0.75/h. 

(2) Yeah, not sure if this favorable feature of selecting CPU/GPU nodes is supported. 

(3) Tried other option to avoid GPU driver like node-label https://docs.aws.amazon.com/emr/latest/ManagementGuide/emr-master-core-task-nodes.html. Did not get it to work.

**Review by eordentlich — APPROVED** [review]
Nice work!

File: notebooks/aws-emr/init-configurations.json
Comment (lijinf2) [reply notebooks/aws-emr/init-configurations.json]: Good question. The CUPY_CACHE_DIR comes from https://docs.cupy.dev/en/stable/reference/environment.html. 
Not sure if CUPY official handles concurrent cupy tasks. Need to try and find out. Created a ticket.

File: notebooks/aws-emr/init-configurations.json
Comment (lijinf2) [reply notebooks/aws-emr/init-configurations.json]: Replaced all tab with four space using auto alignment vim command 'gg=G'.

**Pull request merged**
